---
title: "Introduction to MagenSingleCell"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Introduction to MagenSingleCell}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  results = "hide",
  message = F,
  warning = F
)
library(MagenSingleCell)
```


We used two replicates of CD44<sup>+</sup> T cell data sets from Ciucci et al. 2018 [^1] as an example to demonstrate the use of `MagenSingleCell`. The analysis requires at least 8G of memory on *slurm* [^2] high performance computing workload manager.  We first download the raw 10X data from GEO using `GEOquery`, which can be obtained using the following command if not already installed:

```r
source("https://bioconductor.org/biocLite.R")
biocLite("GEOquery")
```

The two datasets `LCMV1`, `LCMV2` will be downloaded into `~/LCMV`. Each folder will contain the `matrix.mtx`, `gene.tsv`, `barcode.tsv` files as in 10X genomics format.

```{r download_data}
library(MagenSingleCell)
download_LCMV(base_dir = "~/LCMV")
```

We cluster each dataset separately to account for dataset-specific technical and biological differences. Then, we measure the transcriptional similarity and divergence between the clusters identified in the two datasets using correlation analysis.

## Individual analysis of LCMV1 and LCMV2

First, we set up the directory where the results of the analysis will be stored. 

```{r initialize}
LCMV1 <- initialize.project(datasets = "LCMV1", 
                          origins = "CD44+ cells",
                          experiments = "Rep1",
                          data.path = "~/LCMV/",
                          work.path = "~/LCMV/LCMV_analysis")
```

`read.data` function reads the data in 10X genomics format and performs quality filtering as described in *Magen et al 2018* [^3]. We randomly downsampled the datasets to 1000 cells to shorten the running time in this example.

```{r read_LCMV1}
LCMV1 <- read.data(LCMV1, subsample = 1000)
```

Next, we identify highly variable genes on which we base the PCA and clustering analyses. We also compute the activation of gene sets of interest, such as cell cycle genes, which can be used for confounder correction using the `regress` parameter in the PCA analysis.

```{r preprocess}
LCMV1 <- get.variable.genes(LCMV1) 
exhaustion_markers <- c('Pdcd1', 'Cd244', 'Havcr2', 'Ctla4', 'Cd160', 'Lag3', 
                        'Tigit', 'Cd96')
LCMV1 <- add.confounder.variables (LCMV1,
	ribosomal.score = ribosomal.score(LCMV1),
	mitochondrial.score = mitochondrial.score(LCMV1),
	cell.cycle.score = cell.cycle.score(LCMV1),
	Exhaustion = controlled.mean.score(LCMV1, exhaustion_markers))
```

`PCA` function performed PCA after a simulation analysis of shuffled data to determine the appropriate number of PCs. 

```{r PCA}
LCMV1 <- PCA(LCMV1)
```

We perform clustering analysis for a range of clustering resolutions. The analysis is repeated multiple times over shuffled data to estimate the appropriate clustering resolution and control for false discovery of clusters, as described in *Magen et al 2018* [^3].

```{r cluster}
LCMV1 <- cluster.analysis(LCMV1)
```

The clusters are then annotated by customized set of marker genes which the users should adapt to their own data. `summarize` function performs differential expression analysis, computes tSNE and visualizes the results in the analysis folder. 

```{r annotation}
types = rbind(
                data.frame(type='Tfh',gene=c('Tcf7','Cxcr5','Bcl6')),
                data.frame(type='Th1',gene=c('Cxcr6','Ifng','Tbx21')),
                data.frame(type='Tcmp',gene=c('Ccr7','Bcl2','Tcf7')),
                data.frame(type='Treg',gene=c('Foxp3','Il2ra')),
                data.frame(type='Tmem',gene=c('Il7r','Ccr7')),
                data.frame(type='CD8',gene=c('Cd8a')),
                data.frame(type='Cycle',gene=c('Mki67','Top2a','Birc5'))
)
LCMV1_cluster_names <- get.cluster.names(LCMV1, types, min.fold = 1.5, max.Qval = 1e-3)
LCMV1 <- set.cluster.names(LCMV1, names = LCMV1_cluster_names)
summarize(LCMV1)
```

We plot a heatmap containing selected marker genes of interest for initial evaluation.

```{r plot_LCMV1, include = F}
canonical_genes <- c("Cd8a", "Cd4", "Mki67", "Foxp3", "Il2ra", "Bcl6",
                     "Cxcr5", "Cxcr6", "Ifng", "Tbx21", "Id2", "Rora",
                     "Cxcr3", "Tcf7", "Ccr7", "Cxcr4", "Pdcd1", "Ctla4")
MagenSingleCell:::plot.simple.heatmap(LCMV1, name = "canonical", markers = canonical_genes)
```

We repeat the same procedure for `LCMV2` dataset.

```{r LCMV_2}
LCMV2 <- initialize.project(datasets = "LCMV2",
                          origins = "CD44+ cells",
                          experiments = "Rep2",
                          data.path = "~/LCMV/",
                          work.path = "~/LCMV/LCMV_analysis")
LCMV2 <- read.data(LCMV2, subsample = 1000)
LCMV2 <- get.variable.genes(LCMV2)
LCMV2 <- add.confounder.variables(
  LCMV2, 
  ribosomal.score = ribosomal.score(LCMV2),
  mitochondrial.score = mitochondrial.score(LCMV2),
  cell.cycle.score = cell.cycle.score(LCMV2),
  Exhaustion = controlled.mean.score(LCMV2, exhaustion_markers))

LCMV2 <- PCA(LCMV2)
LCMV2 <- cluster.analysis(LCMV2)
LCMV2_cluster_names <- get.cluster.names(LCMV2, types, min.fold = 1.5, max.Qval = 1e-3)
LCMV2 <- set.cluster.names(LCMV2, names = LCMV2_cluster_names)
summarize(LCMV2)

{r plot_LCMV2, include = F}
MagenSingleCell:::plot.simple.heatmap(LCMV2, name = "canonical", markers = canonical_genes)
```

We then initialize the correlation analysis of the two independent runs, providing the information of which analyses folders should be used to pull the data for integration.

```{r initialize_pooled}
pooled_env <- initialize.project(datasets = c("LCMV1", "LCMV2"),
                          origins = rep("CD44+ cells", 2),
                          experiments = c("Rep1", "Rep2"),
                          data.path = "~/LCMV/",
                          work.path = "~/LCMV/LCMV_analysis")
pooled_env <- read.preclustered.datasets(pooled_env)
pooled_env <- add.confounder.variables(
  pooled_env, 
  ribosomal.score = ribosomal.score(pooled_env),
  mitochondrial.score = mitochondrial.score(pooled_env),
  cell.cycle.score = cell.cycle.score(pooled_env),
  Exhaustion = controlled.mean.score(pooled_env, exhaustion_markers))
pooled_env <- PCA(pooled_env, clear.previously.calculated.clustering = F)
summarize(pooled_env, contrast = "datasets")
```

We assessed the similarity between pairs of clusters as described in *Magen et al 2018* [^3] and identify reproducible clusters across the two replicates.

```{r pooled}
cluster.similarity <- compare.cluster.similarity(pooled_env)
similarity <- cluster.similarity$similarity
map <- cluster.similarity$map
filtered.similarity <- get.robust.cluster.similarity(
  pooled_env, similarity, min.sd = qnorm(.9), max.q.val = 0.01, rerun = F
  )
robust.clusters <- sort(unique(c(filtered.similarity$cluster1,
                                 filtered.similarity$cluster2)))
visualize.cluster.cors.heatmaps(pooled_env, pooled_env$work.path,
                                filtered.similarity)
```

Finally, the Pearson correlations between clusters are visualized as a heatmap ordered by hierarchical clustering.

```{r summary}
similarity <- filtered.similarity
visualize.cluster.similarity.stats(pooled_env, similarity)
```

```{bash include = F}
bash ../data-raw/extract_pages.sh
```

### Visualizations 

We used `LCMV2` results to show the visualization output from indiviual analysis from the pipeline. 


<figure>
  <img src="figs/pre_stats.pdf"> 
  <img src="figs/post_stats.pdf">
<figcaption> Fig 1. Mitochondrial score vs UMI for pre (left) and post (right) filtered cells in LCMV2. </figcaption>
</figure>


<figure>
<img src="figs/PCA.pdf"> 
<img src="figs/rotation.pdf">
<figcaption> Fig 2. First two PCs of LCMV2 dataset and corresponding rotation </figcaption>
</figure>


<figure>
<img src="figs/PC1_heatmap.pdf">
<img src="figs/PC2_heatmap.pdf">
<figcaption> Fig 3. Heatmap visualization of the loadings of the first two PCs </figcaption>
</figure>


<figure>
<img src="figs/cluster.size.pdf">
<figcaption> Fig 4. Size of clusters </figcaption>
</figure>

<figure>
<img src="figs/violin.pdf"> 
<figcaption> Fig 5. Cell cycle score violin plot </figcaption>
</figure>

<figure>
<img src="figs/tsne.pdf"> 
<figcaption> Fig 6. t-SNE plot colored by clusters </figcaption>
</figure>


<figure>
<img src="figs/diff.genes.pdf"> 
<figcaption> Fig 7. Heatmap of the canonical markers for T cell subsets </figcaption>
</figure>

Plots for pooled analysis

<figure>
<img src="figs/LCMV1_LCMV2_similarity.pdf"> 
<figcaption> Fig 8. Correlation between pairs of clusters in the two datasets. </figcaption>
</figure>

<figure>
<img src="figs/cor_FC.pdf" width = "600"> 
<figcaption> Fig 9. Correlation among all the clusters in the two datasets </figcaption>
</figure>


[^1]: Ciucci, Thomas, et al. "The Emergence and Functional Fitness of Memory CD4+ T Cells Require the Transcription Factor Thpok." _Immunity_ 50.1 (2019): 91-105.

[^2]: [slurm](https://slurm.schedmd.com)

[^3]: Magen *et al*. “Single-cell profiling of tumor-reactive CD4<sup>+</sup> T-cells reveals unexpected transcriptomic diversity” [*bioRxiv 543199*](https://doi.org/10.1101/543199)
