---
title: "Introduction to MagenSingleCell"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Introduction to MagenSingleCell}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  results = "hide",
  message = F,
  warning = F,
  
)
```


We used two replicates of CD44+ T cell data sets from Ciucci 2018 [^1] as an example to demonstrate the use of `MagenSingleCell`. The analysis requires at least 8G of memory on slurm.  We first download the raw 10X data from GEO using `GEOquery`, which can be obtained using the following command if not already installed:

```r
source("https://bioconductor.org/biocLite.R")
biocLite("GEOquery")
```

The two datasets `LCMV1`, `LCMV2` will be downloaded into `~/LCMV`. Each folder will contain the `matrix.mtx`, `gene.tsv`, `barcode.tsv` files as in 10X genomics format.

```{r download_data, eval = F}
library(MagenSingleCell)
MagenSingleCell::downlaod_LCMV(base_dir = "~/LCMV")
```

We will cluster each dataset separately and perform a correlation analysis of the clusters between the two datasets.

## Individual analysis of LCMV1 and LCMV2

First, we set up the directory where the analysis results will be stored. 

```{r initialize}
LCMV1 <- initialize.project(datasets = "LCMV1", 
                          origins = "CD44+ cells",
                          experiments = "Rep1",
                          data.path = "~/LCMV/LCMV1",
                          work.path = "~/LCMV/LCMV_analysis")
```

`read.data` function reads the data in 10X genomics format. We randomly downsampled the datasets to 1000 cells to shorten the running time. In addition, the function performs quality filtering as in *Magen et al 2018* [^2]. 

```{r read_LCMV1}
LCMV1 <- read.data(LCMV1, subsample = 1000)
```

Next, we obtained the highly variable genes. We also compute gene signatures activation for signatures of interest, which are used for confounder correction. 

```{r preprocess}
LCMV1 <- get.variable.genes(LCMV1) 
exhaustion_markers <- c('Pdcd1','Cd244','Havcr2','Ctla4','Cd160','Lag3','Tigit','Cd96')
LCMV1 <- add.confounder.variables (LCMV1,
	ribosomal.score = ribosomal.score(LCMV1),
	mitochondrial.score = mitochondrial.score(LCMV1),
	cell.cycle.score = cell.cycle.score(LCMV1),
	Exhaustion = controlled.mean.score(LCMV1, exhaustion_markers))
```

`PCA` function performed PCA after a simulation analysis of shuffled data to determine the appropriate number of PCs. 

```{r PCA}
LCMV1 <- PCA(LCMV1)
```

We performed clustering analysis for a range of clustering resolutions. 

```{r cluster}
LCMV1 <- cluster.analysis(LCMV1)
```

The clusters are then annotatedby customized set of marker genes. `summarize` function performs differential expression analysis, computes tSNE and visualizes the results in the analysis folder. 

```{r annotation}
types = rbind(
                data.frame(type='Tfh',gene=c('Tcf7','Cxcr5','Bcl6')),
                data.frame(type='Th1',gene=c('Cxcr6','Ifng','Tbx21')),
                data.frame(type='Tcmp',gene=c('Ccr7','Bcl2','Tcf7')),
                data.frame(type='Treg',gene=c('Foxp3','Il2ra')),
                data.frame(type='Tmem',gene=c('Il7r','Ccr7')),
                data.frame(type='CD8',gene=c('Cd8a')),
                data.frame(type='Cycle',gene=c('Mki67','Top2a','Birc5'))
)
LCMV1_cluster_names = get.cluster.names(LCMV1, types)
LCMV1 <- set.cluster.names(LCMV1, names = LCMV1_cluster_names)
summarize(LCMV1)
```

We repeat the same procedure for `LCMV2` dataset.

```{r LCMV_2}
LCMV2 <- initialize.project(datasets = "LCMV2",
                          origins = "CD44+ cells",
                          experiments = "Rep2",
                          data.path = "~/LCMV/LCMV2",
                          work.path = "~/LCMV/LCMV_analysis")
LCMV2 <- read.data(LCMV2, subsample = 1000)
LCMV2 <- get.variable.genes(LCMV2)
LCMV2 <- add.confounder.variables (LCMV2, ribosomal.score = ribosomal.score(LCMV2),
  mitochondrial.score = mitochondrial.score(LCMV2),
  cell.cycle.score = cell.cycle.score(LCMV2),
  Exhaustion = controlled.mean.score(LCMV2, exhaustion_markers))

LCMV2 <- PCA(LCMV2)
LCMV2 <- cluster.analysis(LCMV2)
LCMV2_cluster_names <- get.cluster.names(LCMV2, types)
LCMV2 <- set.cluster.names(LCMV2, names = LCMV2_cluster_names)
summarize(LCMV2)
```

### Visualizations 


<figure>
<img src="figs/pre_stats.pdf" width="300"> 
<img src="figs/post_stats.pdf" width="300">
<figcaption> Mitochondrial score vs UMI for pre (left) and post (right) filtered cells in LCMV2. </figcaption>
</figure>

<figure>
<img src="figs/PCA.pdf" width="300"> 
<img src="figs/rotations.pdf" width="300">
<figcaption>  </figcaption>
</figure>


[^1]: Ciucci, Thomas, et al. "The Emergence and Functional Fitness of Memory CD4+ T Cells Require the Transcription Factor Thpok." _Immunity_ 50.1 (2019): 91-105.


[^2]: Magen *et al*. “Single-cell profiling of tumor-reactive CD4<sup>+</sup> T-cells reveals unexpected transcriptomic diversity” [*bioRxiv 543199*](https://doi.org/10.1101/543199)