---
title: "LMCV demo"
author: "Meng Wang"
output:
   github_document
always_allow_html: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Data set

If `GEOquery` package is not installed, use `biocLite()` to install the package:

```{r install_GEOquery}
source("http://bioconductor.org/biocLite.R")
biocLite("GEOquery")
```

The example data set comes from Ciucci 2018[^1]. Two replicates of CD44+ cells were available from [GSM3423794](https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSM3423794) and [GSM3423795](https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSM3423795). 


```{r}
dir.create("~/Downloads/LMCV")
rep1_path <- GEOquery::getGEOSuppFiles("GSM3423794", baseDir = "~/Downloads/LMCV")
rep2_path <- GEOquery::getGEOSuppFiles("GSM3423795", baseDir = "~/Downloads/LMCV")
```


```{r initialize}
library(MagenSingleCell)
Exhaustion <- MagenSingleCell:::capwords(c('Pdcd1','Cd244','Havcr2','Ctla4','Cd160','Lag3','Tigit','Cd96'))
env <- initialize.project(datasets = c("GSM3423794", "GSM3423795"), 
                          origins = rep("CD44+ cells", 2),
                          experiments = c("Rep1", "Rep2")
                          data.path = "~/Downloads/LMCV",
                          work.path = "~/Downloads/LMCV_analysis",
                          marker.genes = capwords(unique(marker_genes$symbol)))
```


## Read data

The `read.data` function reads the data in 10X genomics format (`matrix.mtx`, `gene.tsv`, `barcode.tsv`)

```{r}
env <- read.data(env)
```

## Preprocess data

```{r}
env <- get.variable.genes(env) 
env <- add.confounder.variables (env,
	ribosomal.score = ribosomal.score(env),
	mitochondrial.score = mitochondrial.score(env),
	cell.cycle.score = cell.cycle.score(env),
	Exhaustion = controlled.mean.score(env, Exhaustion))
env <- MagenSingleCell::PCA(env)
```

## clustering

```{r}
env <- cluster.analysis(env)
```

## Differential expression and visualizations

[^1] Ciucci, Thomas, et al. "The Emergence and Functional Fitness of Memory CD4+ T Cells Require the Transcription Factor Thpok." _Immunity_ 50.1 (2019): 91-105.
